game-core/src/main/java/games/strategy/triplea/ui/HeadlessUiContext.java

```diff
@@ -68,6 +69,11 @@ public ResourceImageFactory getResourceImageFactory() {
     return null;
   }
 
+  @Override
+  public TerritoryEffectImageFactory getTerritoryEffectImageFactory() {
+    return null;
```

Comment:
@Override returning null usaully indicates an abstraction problem. It's not much of a surprise given headless vs headed was done in a massive hurry to get the bots up and running. In this case it seems like `getTerritoryEffectImageFactory` should only be a method on the `HeadedContext`, not the headless. If a caller is not aware of the difference, they can get a NPE. Seems like a caller of this code would need to know which instance type, so moving the method to the right instance type I suspect is appropriate and would avoid the error and abstraction problem.

Reply:
If the caller doesn't know it's running in a headed context, another alternative to get rid of the `null` smell is to make `TerritoryEffectImageFactory` an interface and have `Headed` and `Headless` implementations, where the latter is effectively a _Null Object_.

Reply:
Yeah, its probably not the best pattern but I just followed how the existing image factories were used.

Reply:
That's not very compelling, instead of fixing the jenga tower, another block was added.

Unless I missed something, this method is used in exactly one place:
```java              territoryEffectLabel.setIcon(uiContext.getTerritoryEffectImageFactory().getIcon(territoryEffect, false));
```
(TripleAFrame).

We will get a NPE if the wrong instance is there. We can do a bit better by casting and then we avoid the needless additional API on HeadlessUiContext:

```
territoryEffectLabel.setIcon((HeadedUiContext) uiContext).getTerritoryEffectImageFactory().getIcon(territoryEffect, false));
```

It's a very simple change, does not solve the abstraction fully, but we do not make it worse.

Reply:
I'd vote to keep consistency here as I think either way its still not great. If someone was to come back to this code later and see ResourceImageFactory have one pattern and TerritoryEffectsImageFactory have another pattern it would probably be confusing. Probably better to consider doing a refactor of the HeadlessUiContext as a whole to determine the best way to remove the numerous methods returning null.